### ISSUE TITLE
Support for IDN

### ISSUE THREAD

**Issue by ad-m** [octodns/octodns#199]
Hello,

I am owner of ```zajęzyk.pl``` (punycode: ```xn--zajzyk-y4a.pl```) hosted on CloudFlare.

I got following errors:
```
$ octodns-dump --config-file=env/production.yaml --output-dir=zones zajęzyk.pl. cloudflare_shared
2018-02-18T12:58:33  [140367053055744] INFO  Manager __init__: config_file=env/production.yaml
2018-02-18T12:58:33  [140367053055744] INFO  Manager __init__:   max_workers=1
2018-02-18T12:58:33  [140367053055744] INFO  Manager __init__:   max_workers=False
Traceback (most recent call last):
  File "/usr/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc4 in position 3: ordinal not in range(128)
Logged from file manager.py, line 347
Traceback (most recent call last):
  File "/home/adas/.virtualenvs/infra/bin/octodns-dump", line 11, in <module>
    sys.exit(main())
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/cmds/dump.py", line 31, in main
    manager.dump(args.zone, args.output_dir, args.lenient, *args.source)
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/manager.py", line 360, in dump
    zone = Zone(zone, self.configured_sub_zones(zone))
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/manager.py", line 159, in configured_sub_zones
    pieces = zone_name[:-1].split('.')[::-1]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc4 in position 3: ordinal not in range(128)
$ octodns-dump --config-file=env/production.yaml --output-dir=zones xn--zajzyk-y4a.pl. cloudflare_shared
2018-02-18T12:59:28  [139790821807872] INFO  Manager __init__: config_file=env/production.yaml
2018-02-18T12:59:28  [139790821807872] INFO  Manager __init__:   max_workers=1
2018-02-18T12:59:28  [139790821807872] INFO  Manager __init__:   max_workers=False
2018-02-18T12:59:28  [139790821807872] INFO  Manager dump: zone=xn--zajzyk-y4a.pl., sources=()
2018-02-18T12:59:29  [139790821807872] INFO  CloudflareProvider[cloudflare_shared] populate:   found 0 records
2018-02-18T12:59:29  [139790821807872] INFO  YamlProvider[dump] plan: desired=xn--zajzyk-y4a.pl.
2018-02-18T12:59:29  [139790821807872] INFO  YamlProvider[dump] plan:   No changes
2018-02-18T12:59:29  [139790821807872] INFO  YamlProvider[dump] apply: making changes
```

No DNS records found for punycode. There is records for that domain:

![zaznaczenie_1353](https://user-images.githubusercontent.com/3618479/36351524-b34ee3b6-14ab-11e8-8c84-314c09ab16b2.png)

I would like to work with IDN domains just like with regular domains. It is indifferent whether through direct or indirect, but the export must be.

```
$ python --version
Python 2.7.12
$ pip freeze | grep 'octodns'
octodns==0.8.8
```

**Comment by ross** [comment]
Started poking around at this one and fro the looks of it the fixes are going to be a bit involved as the Cloudflare provider is only one of the places where problems arise. In general a lot of the string handling stuff is blowing up. Early guess is that some of the problems I'm seeing are coming from the YAML library, but I'm not yet 100% sure about that. 

This is closely related to #201 re python3 support as string handling will be one of the places most likely to need cleanup and adjustments.

**Comment by ross** [comment]
I have a PR in the works that will at least get the octoDNS internals able to work with non-ascii chars (mostly using `unicode` instead of `str` at least for python2. That said all of the providers have problems working with the records (I copied your example from above). Cloudflare is the only one where I can get an initial sync to work. It does create the records and a subsequent sync doesn't see any diffs. I run into problems thought when I try and make changes:

```
(env)otter:octodns ross$ PYTHONPATH=. ./octodns/cmds/sync.py --config-file=config/dev.yaml --doit
2018-02-24T08:37:17  [140736612635584] INFO  Manager __init__: config_file=config/dev.yaml
2018-02-24T08:37:17  [140736612635584] INFO  Manager __init__:   max_workers=4
2018-02-24T08:37:17  [140736612635584] INFO  Manager __init__:   max_workers=True
2018-02-24T08:37:17  [140736612635584] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=False, force=False
2018-02-24T08:37:17  [140736612635584] INFO  Manager sync:   zone=zajęzyk.pl.
2018-02-24T08:37:17  [140736612635584] INFO  Manager sync:   sources=['config'] -> targets=['cloudflare']
2018-02-24T08:37:17  [123145309708288] INFO  YamlProvider[config] populate:   found 2 records
2018-02-24T08:37:17  [123145309708288] INFO  CloudflareProvider[cloudflare] plan: desired=zajęzyk.pl.
2018-02-24T08:37:18  [123145309708288] INFO  CloudflareProvider[cloudflare] populate:   found 3 records
2018-02-24T08:37:18  [123145309708288] INFO  CloudflareProvider[cloudflare] plan:   filtered out 1 changes
2018-02-24T08:37:18  [123145309708288] INFO  CloudflareProvider[cloudflare] plan:   Creates=0, Updates=1, Deletes=0, Existing Records=3
2018-02-24T08:37:18  [140736612635584] INFO  Manager
********************************************************************************
* zajęzyk.pl.
********************************************************************************
* cloudflare (CloudflareProvider)
*   Update
*     <AliasRecord ALIAS 3600, zajęzyk.pl., s7.zenbox.pl.> ->
*     <AliasRecord ALIAS 3600, zajęzyk.pl., s7.zenbox2.pl.> (config)
*   Summary: Creates=0, Updates=1, Deletes=0, Existing Records=3
********************************************************************************


2018-02-24T08:37:18  [140736612635584] INFO  CloudflareProvider[cloudflare] apply: making changes
Traceback (most recent call last):
  File "./octodns/cmds/sync.py", line 43, in <module>
    main()
  File "./octodns/cmds/sync.py", line 39, in main
    dry_run=not args.doit, force=args.force)
  File "/Users/ross/github/octodns/octodns/manager.py", line 311, in sync
    total_changes += target.apply(plan)
  File "/Users/ross/github/octodns/octodns/provider/base.py", line 87, in apply
    self._apply(plan)
  File "/Users/ross/github/octodns/octodns/provider/cloudflare.py", line 456, in _apply
    getattr(self, '_apply_{}'.format(class_name))(change)
  File "/Users/ross/github/octodns/octodns/provider/cloudflare.py", line 424, in _apply_Update
    self._request('POST', path, data=content)
  File "/Users/ross/github/octodns/octodns/provider/cloudflare.py", line 80, in _request
    raise CloudflareError(resp.json())
octodns.provider.cloudflare.CloudflareError: An A, AAAA or CNAME record already exists with that host.
```

I'm still looking to see if this is a Cloudflare problem or an issue with the provider. 

Every other provider option I've tried fails in either the client library or when the API call is made. So the TL;DR is that I'm going to get octoDNS in a place where it doesn't stupidly barf for non-ascii stuff and I might be able to get Cloudflare working, but otherwise more general support will likely have to be done by someone with more knolwedge of the details of the subject and the time to take it on that I have.

**Comment by ross** [comment]
Figured out what was up. Unrelated problem that I've addressed in https://github.com/github/octodns/pull/208. 

https://github.com/github/octodns/pull/209 should get Cloudflare working with non-ascii stuff.

**Comment by ad-m** [comment]
There is still some issue.

```
$ octodns-dump --config-file=env/production.yaml --output-dir=tmp/ zajęzyk.pl cloudflare_shared
2018-02-25T19:13:37  [139825236637440] INFO  Manager __init__: config_file=env/production.yaml
2018-02-25T19:13:37  [139825236637440] INFO  Manager __init__:   max_workers=1
2018-02-25T19:13:37  [139825236637440] INFO  Manager __init__:   max_workers=False
Traceback (most recent call last):
  File "/usr/lib/python2.7/logging/__init__.py", line 861, in emit
    msg = self.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/usr/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc4 in position 3: ordinal not in range(128)
Logged from file manager.py, line 345
Traceback (most recent call last):
  File "/home/adas/.virtualenvs/infra/bin/octodns-dump", line 11, in <module>
    load_entry_point('octodns==0.8.8', 'console_scripts', 'octodns-dump')()
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/cmds/dump.py", line 31, in main
    manager.dump(args.zone, args.output_dir, args.lenient, *args.source)
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/manager.py", line 358, in dump
    zone = Zone(zone, self.configured_sub_zones(zone))
  File "/home/adas/.virtualenvs/infra/local/lib/python2.7/site-packages/octodns/manager.py", line 188, in configured_sub_zones
    pieces = zone_name[:-1].split('.')[::-1]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc4 in position 3: ordinal not in range(128)
```

**Comment by ross** [comment]
Will take a look at the dump flow the next time I get a chance to dig into things.

**Comment by tnir** [comment]
@ross Yes, I don't think #209 enabled us to support IDN, but also the PR made it worse for Python 3 support.

`.encode('idna')` will help us support IDN. Or can we use https://pypi.org/project/idna/ ?

**Comment by PeterDaveHello** [comment]
If IDN is not supported yet, should we make sure it's invalid and failed when octodns processing such records? In case the other users thought it should work?

**Comment by ross** [comment]
Ideally it'd work, but I don't have a use case for it and haven't had the bandwidth to look at supporting it. If someone else does that'd be awesome. Re rejecting them if support isn't there, kind of the same place in terms of I'm not going to have the bandwidth to implement that, but if someone else did we could explore it.

**Comment by PeterDaveHello** [comment]
I might be able to help with the rejecting part before it's supported. Maybe after #628 finished 😄

**Comment by PeterDaveHello** [comment]
cc https://github.com/ypcrts/fqdn/issues/25

**Comment by PeterDaveHello** [comment]
Should keep it open.

**Comment by tnir** [comment]
Do not close this issue 👍

**Comment by ross** [comment]
> `.encode('idna')` will help us support IDN. Or can we use https://pypi.org/project/idna/ ?

Dug into this a bit and the TL;DR is that the `idna` module probably is the right way to go, but there isn't a clean and simple place to plug it in and get things to just work. If `Zone` and `Record` encode the names of things when they're created it causes them to show in their punycode everywhere and even try and load incorrect filenames etc. 

Now that the linked PRs have shipped and octodns is fully into pthon3/utf8 the core library and `YamlProvider` are OK dealing with them, but most/all providers fail when they go to use utf8 chars to create zones or records:

```yaml
(env) coho:octodns ross$ PYTHONPATH=. ./octodns/cmds/sync.py --config-file=config/dev.yaml
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__: config_file=config/dev.yaml (octoDNS 0.9.17+eccf91f3)
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__:   max_workers=1
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__:   include_meta=False
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.17+eccf91f3)
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__: provider=ns1 (octodns_ns1 0.0.2)
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__: provider=route53 (octodns_route53 0.0.4)
2022-05-20T07:59:50  [4544204288] INFO  Manager __init__: processor=acme (octodns.processor.acme 0.9.17+eccf91f3)
2022-05-20T07:59:50  [4544204288] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=True, force=False, plan_output_fh=<stdout>
2022-05-20T07:59:50  [4544204288] INFO  Manager sync:   zone=zajęzyk.pl.
2022-05-20T07:59:50  [4544204288] INFO  Manager sync:   sources=['config'] -> targets=['route53']
2022-05-20T07:59:50  [4544204288] INFO  YamlProvider[config] populate:   found 2 records, exists=False
2022-05-20T07:59:50  [4544204288] INFO  Route53Provider[route53] plan: desired=zajęzyk.pl.
2022-05-20T07:59:51  [4544204288] INFO  Route53Provider[route53] populate:   found 0 records, exists=False
2022-05-20T07:59:51  [4544204288] INFO  Route53Provider[route53] plan:   Creates=2, Updates=0, Deletes=0, Existing Records=0
2022-05-20T07:59:51  [4544204288] INFO  Manager
********************************************************************************
* zajęzyk.pl.
********************************************************************************
* route53 (Route53Provider)
*   Create Zone<zajęzyk.pl.>
*   Create <ARecord A 9, *.zajęzyk.pl., ['1.2.3.4']> (config)
*   Create <ARecord A 9, zajęzyk.zajęzyk.pl., ['2.3.4.5']> (config)
*   Summary: Creates=2, Updates=0, Deletes=0, Existing Records=0
********************************************************************************
```

I'm going to keep digging, but I'm a little concerned that this will be something that each provider will need to deal with, potentially with `zone.idna_name` and `record.idna_name` helpers, but still a fairly involved change.

**Comment by ross** [comment]
Created https://github.com/octodns/octodns/pull/903 which adds some idna helpers to the core libary, otherwise octodns core itself, including `YamlProvider`, is good. Looks like the idna module doesn't handle `*` and `_` in names so that's something the helpers are going to have to deal with.

I then started in on https://github.com/octodns/octodns-ns1/pull/20 using that new functionality. It's still WIP and it especially needs some testing around dynamic record support. Unfortunately this is likely going to require updates to all of the provider modules individually 😦

**Comment by ross** [comment]
octoDNS itself now has the IDN helpers required to implement this support and the remaining work will happen in each provider. They'll be linked up to this issue as that happens. I'll probably only look at doing the ones that I have access to and thus can verify. Others should feel free to submit PRs. They'll likely be fairly simple PRs, especially for providers w/o dynamic support. 

Going to call this closed since octoDNS itself is done.

**Comment by ross** [comment]
Reopening based deeper digging. The more I do so the messier this gets. I'm realizing now that if a provider supports IDNA and a user has configured a manually translated zone (as is required prior to octoDNS IDNA support) then the changes like this will break their setup as they'll have xn--... zones and records in the config and when the provider pulls that data down during populate it'll be translating it to the utf-8 equivilent and the two zones being compared won't even match.

I'm abandoning this and all of the existing idna PRs for the moment as I feel like the path things are going down will not be viable. Good news is the ida_encode/decode shipped in octoDNS still makes sense and nothing else has actually be part of a merged PR/release so we should be in an ok state until things are figured out.

Rethinking options for how this might work. We'll see how that goes if I come up with something viable beyond people having to manually translate things.

Notes:

Things could be translated as loaded by YamlProvider, but the downside of that is everything that's show as output during sync would be the IDNA versions which would be complicated for everyone and unusable for non-technical folks. 

Similar problem would be the case if things automatically encoded/decoded at the Record level as they were created. E.g. if `fqdn` always spit out the `IDNA` version providers would automatically work which would be nice, but the output to users wouldn't match what they expected/was on disk.

Anyway, I'd like to get this done, but it's getting more and more involved the more I look into is to we'll see.

**Comment by ross** [comment]
Nth time is maybe the charm, https://github.com/octodns/octodns/pull/922

TL;DR is that I struck on the idea of representing everything internally with idna encoded (that means as-is for ascii.) That coupled with creating an IdnaDict that is insensitive to case and idna encoding makes everything just work:tm:. More info in the PR. Thoughts solicited.

**Comment by ross** [comment]
Addressed in https://github.com/octodns/octodns/pull/922
